x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java

```diff
@@ -233,9 +233,46 @@ public void testListenerCalledImmediatelyWhenProgressReached() {
         assertThat(future.isDone(), is(true));
     }
 
+    public void testLongConsumerCalledOnProgressUpdate() {
+        // min length of 2 to have at least one progress update before reaching the end
+        long length = randomLongBetween(2L, ByteSizeUnit.TB.toBytes(1L));
+        long start = randomLongBetween(Long.MIN_VALUE, Long.MAX_VALUE - length);
+        long end = start + length;
+
+        var consumed = new HashSet<Long>();
+        var future = new ProgressListenableActionFuture(
+            start,
+            end,
+            p -> assertThat("LongConsumer should not consumed the same  value twice", consumed.add(p), equalTo(true))
+        );
+
+        long position = start;
+        for (int i = 0; i < 25 && position < end - 1L; i++) {
+            var progress = randomLongBetween(position + 1L, end - 1L);
+
+            var listener = new PlainActionFuture<Long>();
+            future.addListener(
+                ActionListener.runBefore(
+                    listener,
+                    () -> assertThat(
+                        "LongConsumer should have been called before listener completion",
+                        consumed.contains(progress),
+                        equalTo(true)
+                    )
+                ),
+                randomLongBetween(position + 1L, progress)
+            );
+            future.onProgress(progress);
+
+            assertThat(consumed.contains(progress), equalTo(true));
+            assertThat(listener.isDone(), equalTo(true));
+            position = progress;
+        }
```

Comment:
The consumer is not called for `future.onProgress(end)`. Is there any value to assert that? Or do you think this is an implementation detail and not worth it?

Reply:
I added a check that the consumer is not called when progress is updated to `end` in 16052d4fa5cf003d0243518f41c08c51a8002314